### ISSUE TITLE
Alpine Linux package ideas and questions

### ISSUE THREAD

**Issue by andyshinn** [firehol/netdata#168]
This is related to #66 and #77. Basically, looking for a couple things on the packaging side and some testing before pushing upstream. Package at https://github.com/andyshinn/alpine-pkg-netdata (downloaded apk in releases).

Some questions that came up during building:
- The default package has no default `netdata.conf`. It doesn't exit. But does give an ERROR that the config is missing. If I just `touch` the file the error goes away. Is there a better way to handle a missing or default configuration file in this case?
- Some stuff in `/usr/share/netdata/web` seems to get written as `root` on first run and causes the interface to complain. Is this something we can solve outside of the package so we don't need to do some sort of `chown` in a startup script?
- In no daemon mode, it doesn't seem to be responding to control signals (can't CTRL-C or send TERM signal). Something we can fix?
- Why is `apps.plugin` setuid? Does it need to be?
- What is `tc` used for? The error log complains about it being missing. We can easily add it with the `iproute2` package. But the usual approach to Alpine packages is usually "minimal first". If it is only needed by a plugin maybe we can disable the plugin by default?

Here is a `Dockerfile` for testing the package:

```
FROM alpine:3.3
RUN apk -U add ca-certificates \
    && wget -q -O /etc/apk/keys/andyshinn.rsa.pub https://raw.githubusercontent.com/andyshinn/alpine-pkg-netdata/master/andyshinn.rsa.pub \
    && wget https://github.com/andyshinn/alpine-pkg-netdata/releases/download/unreleased/netdata-1.0.1-r1.apk \
    && apk add netdata-1.0.1-r1.apk \
    && rm -rf netdata-1.0.1-r1.apk /var/cache/apk/* \
    && touch /etc/netdata/netdata.conf
ENTRYPOINT ["netdata", "-nd"]
```

You can run with something like:

```
docker run -d --name netdata -p 19999:19999 -v /proc:/host/proc:ro -v /sys:/host/sys:ro netdata -ch /host
```

**Comment by titpetric** [comment]
I'd add one more thought to this: automated builds. In order for the build to be automated, it should do a pull from git (when pushed to git). Otherwise, the package would be build from the apk which is available in your releases repo. Can we look at automating this? To be automated now, the apk would have to be built before the git push to the repo - this will never happen :)

I will research best practice a bit on how to provide multiple docker files within one repo, just so we don't get stuck in what's on debian and what's on alpine. Currently debian seems to have less issues, and there's no reason that individual dockerfiles for debian AND alpine shouldn't exist.

**Comment by andyshinn** [comment]
The package will track a proper release version once there is one with the musl fixes. On any version bump we just open a pull request to the mailing list to bump the version in `APKBUILD`. The current references to a specific commit is just so I can test the musl fixes.

**Comment by ktsaou** [comment]
> The default package has no default netdata.conf. It doesn't exit. But does give an ERROR that the config is missing. If I just touch the file the error goes away. Is there a better way to handle a missing or default configuration file in this case?

The configuration file is optional (netdata has internal defaults for everything). Something should be printed if missing though. It is not required to have the whole of it either. For example, when packaging you can place there defaults you see fit for your particular case.

> Some stuff in /usr/share/netdata/web seems to get written as root on first run and causes the interface to complain. Is this something we can solve outside of the package so we don't need to do some sort of chown in a startup script?

Chowning files from the startup script is a bad practice. However, I am not sure this should be done by the makefiles (users build the software are root and as normal users - in the the latter case a chown will mess things totally). 

Ideally you could have a script at installation to do it (this is what `netdata-installer.sh` does).

> In no daemon mode, it doesn't seem to be responding to control signals (can't CTRL-C or send TERM signal). Something we can fix?

HUP is generated also when a plugin stops. If we respond to it, netdata will also exit when one of its childs exits. Also, `stdout` and `stderr` are sent to files, so the console file is closed.

INT (control-C) I see that it should be handled. I don't know if the stdout/stderr is responsible for ignoring it. Please open a separate issue and I'll try to find out what is happening.

> Why is apps.plugin setuid? Does it need to be?

Check #157 

> What is tc used for? The error log complains about it being missing. We can easily add it with the iproute2 package. But the usual approach to Alpine packages is usually "minimal first". If it is only needed by a plugin maybe we can disable the plugin by default?

`tc` is Linux QoS (check https://github.com/firehol/firehol/wiki/FireQOS-Tutorial - also a software piece of the firehol team).

**Comment by andyshinn** [comment]
> Chowning files from the startup script is a bad practice. However, I am not sure this should be done by > the makefiles (users build the software are root and as normal users - in the the latter case a chown will > mess things totally).

The files in question are written as `root` at netstat startup even though the daemon changes user to `netdata`. Shouldn't the files be written as `netdata`?

**Comment by ktsaou** [comment]
> Shouldn't the files be written as netdata

There another long discussion on #186 about the files. There we concluded that the optimal for the web files is:

 - `web files user = root`
 - `web files group = netdata`
 - package files in `/usr/share` as `%defattr(755,root,netdata,644)`

However, I don't think you should do this from a startup script. The user may decide to change these settings. If you change them from the startup script, hardcoded, every time the user restarts the server, netdata will not work.

I suggest to chown/chmod them, only during installation.

**Comment by andyshinn** [comment]
I think I have this package almost working now with 1.1.0. Some more questions:

* I am using the included OpenRC init script. But it doesn't seem to be able to stop the netdata process. I also cannot kill the process with a simple `SIGTERM`. Should this be working? See error log below.
* I had to patch `log.h` to include <time.h> due to error `log.h:45:8: error: unknown type name 'time_t'`. Is this safe to include upstream (probably a musl compatibility thing).

```
16-04-25 00:43:39: INFO: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/charts.d.plugin' on pid 3564 stopped.
16-04-25 00:43:39: ERROR: netdata: pid 3564 exited with code 0. (errno 22, (null))
16-04-25 00:43:39: ERROR: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/charts.d.plugin' (pid 3564) does not generate useful output. Waiting a bit before starting it again. (errno 22, (null))
16-04-25 00:43:40: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:41: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:42: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:43: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:43: INFO: netdata: Signaled exit (signal 15). (errno 22, (null))
16-04-25 00:43:43: ERROR: netdata: Cannot unlink pidfile '/run/netdata.pid'. (errno 22, (null))
16-04-25 00:43:43: INFO: netdata: NetData exiting. Bye bye...
16-04-25 00:43:44: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:44: INFO: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/apps.plugin' on pid 2242 stopped.
16-04-25 00:43:45: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:46: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:47: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:48: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:49: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:50: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:51: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
16-04-25 00:43:52: ERROR: apps.plugin: pid 2232 netdata states parent 1, but the later does not exist. (errno 22, (null))
```

**Comment by ktsaou** [comment]
Could you please post the netdata process tree before and after killing it with SIGTERM, using `ps -fax`?

And apps.plugin complains about pid 1. You don't have process `init` (pid 1)?

Regarding the subdirectory `/run`, the maintainers of the distros did not like it, so it has been removed.

**Comment by ktsaou** [comment]
We are trying to cleanup the issues. Is there anything pending on this issue?

**Comment by andyshinn** [comment]
I built out a package at https://github.com/andyshinn/alpine-pkg-netdata and had two minor patches to get working. I'll bump to 1.2.0 in the next couple days and see how it goes. If everything looks good, i'll send a patch upstream and we can try getting it in Alpine 3.5+.

**Comment by ktsaou** [comment]
perfect! thanks!

**Comment by andyshinn** [comment]
Were xattrs set on `apps.plugin` somewhere? I'm running into this while testing in a container:

```
ERROR: Failed to set xattrs on usr/libexec/netdata/plugins.d/apps.plugin.apk-new: Operation not permitted
```

Specifically, I think this is because my testing environment runs on BTRFS where xattrs are not supported. Not sure if this is an Alpine package side affect or netdata yet. Can't find any evidence that xattrs are being set during make install but thought I'd ask...

**Comment by ktsaou** [comment]
apps.plugin uses capabilities (`setcap`) or if this is not available, it is setuid to root.

**Comment by ktsaou** [comment]
netdata now runs cleanly on alpine and musl.
I have settled all the issues, compilation and run-time.
I plan to release 1.4 this weekend.

**Comment by andyshinn** [comment]
Cool, I will take a look at creating package again against 1.4 tarball.

**Comment by ktsaou** [comment]
Even the installer now fully supports alpine (it is documented at the installation page).
What is missing is an init script. If you have one, please make a PR.

I am closing this now.
